File: projectNew/old_src/TTA/CopyTacticsListener.java
package com.f14.TTA.listener;

import com.f14.TTA.TTAGameMode;
import com.f14.TTA.TTAPlayer;
import com.f14.TTA.component.card.TTACard;
import com.f14.TTA.consts.TTAGameCmd;
import com.f14.bg.action.BgAction;
import com.f14.bg.action.BgResponse;
import com.f14.bg.exception.BoardGameException;
import com.f14.bg.listener.InterruptParam;
import com.f14.bg.utils.BgUtils;
import com.f14.bg.utils.CheckUtils;

import java.util.List;

/**
 * 拷贝阵型的监听器
 *
 * @author F14eagle
 */
public class CopyTacticsListener extends TTAInterruptListener {

    protected TTACard card = null;

    public CopyTacticsListener(TTAGameMode gameMode, TTAPlayer trigPlayer) {
        super(gameMode, trigPlayer);
        this.addListeningPlayer(trigPlayer);
    }


    @Override
    public InterruptParam createInterruptParam() {
        InterruptParam param = new InterruptParam();
        param.set("card", card);
        return param;
    }

    @Override
    protected BgResponse createStartListenCommand(TTAPlayer player) {
        BgResponse res = super.createStartListenCommand(player);
        // 传递当前公共阵型库
        res.setPrivateParameter("multiSelection", false);
        res.setPrivateParameter("cardIds",
                BgUtils.card2String(gameMode.getCardBoard().getPublicTacticsDeck().getCards()));
        return res;
    }

    @Override
    protected void doAction(BgAction action) throws BoardGameException {
        boolean confirm = action.getAsBoolean("confirm");
        TTAPlayer player = action.getPlayer();
        if (confirm) {
            String cardIds = action.getAsString("cardIds");
            List<TTACard> cards = gameMode.getCardBoard().getPublicTacticsDeck().getCards(cardIds);
            CheckUtils.check(cards.size() != 1, "没有找到指定阵型!");
            TTACard card = cards.get(0);
            CheckUtils.check(card == player.getTactics(), "你正在使用这个阵型!");
            this.card = card;
        }
        this.setPlayerResponsed(gameMode, player);
    }


    @Override
    protected String getMsg(TTAPlayer player) {
        return "请选择需要学习的阵型";
    }

    @Override
    protected int getValidCode() {
        return TTAGameCmd.GAME_CODE_COPY_TACTICS;
    }
}

File: projectNew/src/com/f14/TTA/listener/CopyTacticsListener.kt
package com.f14.TTA.listener

import com.f14.TTA.TTAGameMode
import com.f14.TTA.TTAPlayer
import com.f14.TTA.component.card.TTACard
import com.f14.TTA.consts.TTAGameCmd
import com.f14.bg.action.BgAction.GameAction
import com.f14.bg.action.BgResponse
import com.f14.bg.exception.BoardGameException
import com.f14.bg.listener.InterruptParam
import com.f14.bg.utils.BgUtils
import com.f14.bg.utils.CheckUtils

/**
 * Created by 吹风奈奈 on 2017/7/20.
 */

class CopyTacticsListener(gameMode: TTAGameMode, trigPlayer: TTAPlayer) : TTAInterruptListener(gameMode, trigPlayer) {

    private var card: TTACard? = null

    init {
        this.addListeningPlayer(trigPlayer)
    }


    override fun createInterruptParam(): InterruptParam {
        val param = super.createInterruptParam()
        param["card"] = card
        return param
    }

    override fun createStartListenCommand(player: TTAPlayer): BgResponse {
        val res = super.createStartListenCommand(player)
        // 传递当前公共阵型库
        res.private("multiSelection", false)
        res.private("cardIds", BgUtils.card2String(gameMode.cardBoard.publicTacticsDeck.cards))
        return res
    }

    @Throws(BoardGameException::class)
    override fun doAction(action: GameAction) {
        val confirm = action.getAsBoolean("confirm")
        val player = action.getPlayer<TTAPlayer>()
        if (confirm) {
            val cardIds = action.getAsString("cardIds")
            val cards = gameMode.cardBoard.publicTacticsDeck.getCards(cardIds)
            CheckUtils.check(cards.size != 1, "没有找到指定阵型!")
            val card = cards[0]
            CheckUtils.check(card === player.tactics, "你正在使用这个阵型!")
            this.card = card
        }
        this.setPlayerResponsed(player)
    }


    override fun getMsg(player: TTAPlayer): String {
        return "请选择需要学习的阵型"
    }

    override val validCode: Int
        get() = TTAGameCmd.GAME_CODE_COPY_TACTICS
}
